ehsm_kms_service/function.js

```diff
@@ -21,6 +31,172 @@ const base64_encode = (str) => new Buffer.from(str).toString('base64')
+      try {
+        const current_time = new Date().getTime() + CMK_EXPIRE_TIME_EXPAND
+        const query = {
+          selector: {
+            //Query criteria
+            expireTime: { $lt: current_time },
+          },
+          fields: ['_id', '_rev'], // Fields returned after query
+          limit: 10000,
+        }
+        DB.partitionedFind('cmk', query) // Query expired cmks
+          .then((cmks_res) => {
+            for (const cmk_item of cmks_res.docs) {
+              cmk_item._deleted = true
+            }
+            DB.bulk({ docs: cmks_res.docs }) // Batch delete expired cmks
+              .catch((err) => {
+                console.log(err)
+              })
+          })
+          .catch((err) => {
+            console.log(err)
+          })
+      } catch (err) {
+        console.log(err)
+      }
+    }, CMK_LOOP_CLEAR_TIME)
+  }, recent - nowTime)
+
+  return { timer }
+}
+
+/**
+ * @param {object} napi_res
+ * @param {object} res
+ * @param {string} appid
+ * @param {object} DB
+ * @param {object} payload
+ * Fields contained in the cmk document:
+ *    _id | keyid | keyBlob | creator | creationDate | expireTime | alias | keyspec | origin
+ *
+ * After storing cmk successfully, return the keyid to the user
+ */
+function store_cmk(napi_res, res, appid, payload, DB) {
+  try {
+    const creationDate = new Date().getTime()
+    const keyid = uuidv4().replace(/-/g, '')
+    let { keyspec, origin } = payload
+
+    DB.insert({
+      _id: `cmk:${keyid}`,
+      keyid,
+      keyBlob: napi_res.result.cmk_base64,
+      creator: appid,
+      creationDate,
+      expireTime: creationDate + CMK_EFFECTIVE_DURATION,
+      alias: '',
+      keyspec,
+      origin,
+    })
+      .then((r) => {
+        delete napi_res.result.cmk_base64 // Delete cmk_base64 in NaPi result
+        napi_res.result.keyid = keyid // The keyID field is added to the result returned to the user
+        res.send(napi_res)
+      })
+      .catch((e) => {
+        res.send(_result(400, 'create cmk faild', e))
+      })
+  } catch (e) {
+    res.send(_result(400, 'create cmk faild', e))
+  }
+}
+
+/**
+ * ehsm napi result
+ * @param {function name} action
+ * @param {} res
+ * @param {NAPI_* function params} params
+ * @returns
+ */
+function napi_result(action, res, params) {
+  try {
+    const napi_res = ehsm_napi[`NAPI_${action}`](...params)
+    if (JSON.parse(napi_res).code != 200) {
+      res.send(napi_res)
+      return false
+    } else {
+      return JSON.parse(napi_res)
+    }
+    // })
+  } catch (e) {
+    return _result(400, 'Parsing error')
+  }
+}
+
+/**
+ * test create_app_info save in couchDB
+ * @param {object} DB
+ * @param {object} res
+ * Fields contained in the user_info document:
+ * _id | appid | appkey | cmk
+ */
+const create_app_info = (DB, res) => {
+  const appid = new Date().getTime() + ''
+  DB.insert({
+    _id: `user_info:${appid}`,
+    appid: appid,
+    appkey: appid,
+    cmk: '_',

```

Comment:
create_app_info -> create_user_info
These items of user info are not finalized yet, right?
cmk could be multiple.  appid and key cannot use getTime(). And some additional items should be added, e.g. created time 
If it's for test only, it can be merged but please update them in future patches.

Reply:
it will be fix

yes, not finalized, test only

Reply:
OK. you can merge this version and update it in the future.